### PR TITLE
Add c2hs mode

### PR DESCRIPTION
**Pull request by sergv**
This is a wrapper around haskell-mode, that mostly adds fontification for {#...#} blocks.

### CODE REVIEW

File: haskell-c2hs.el [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+;; This program is distributed in the hope that it will be useful,
+;; but WITHOUT ANY WARRANTY; without even the implied warranty of
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+;;; Commentary:
+;; This mode is mostly indended for highlighting {#...#} hooks.
```
Comment (geraldus) [inline haskell-c2hs.el]: `indended` — typo? (:

**Comment by sergv** [comment]
Fixed typo in the description.

**Comment by geraldus** [comment]
Would you like to add some tests?

**Comment by sergv** [comment]
I think I'll add some font-lock tests

**Comment by gracjan** [comment]
Yes, please add some tests. These are required, otherwise we will not be able to take responsibility of this module.

And thanks for great contribution!

**Comment by sergv** [comment]
I added tests for all hooks currently colorized and rebased against latest master

**Comment by gracjan** [comment]
Nice. Lets wait 2 days before merging so that people have time to look and test the code.

**Comment by gracjan** [comment]
Okay, so this merge wasn't the cleanest one and we had issue reports that melpa compiled version did not load.

1. All symbols, including `defface` should start with `haskell-`.
2. `autoloads.el` should actually load, it did not in this case.
3. Overuse of macros is not a good sign.

@sergv: Can you inline the `haskell--let-rx` macro? It overcomplicates things. Please also add `haskell-` prefix to `defface`s. Thanks.

**Comment by gracjan** [comment]
The `autoloads` issue will never happen again thanks to #1211.

**Comment by sergv** [comment]
@gracjan I'll inline the macro and add prefixes, stay tuned!

**Comment by sergv** [comment]
@gracjan why are macros not a good sing? The haskell--let-rx was introduced to make writing regexps via standard rx macro easier and make them both expand at compile time. It's easy to inline, but the advantage of compile time execution will be lost.

**Comment by gracjan** [comment]
Macros are another layer of abstraction that a potential contributor needs to understand. As such it poses a possible mental roadblock which we would rather like to avoid.

**Comment by sergv** [comment]
Fair enough

**Pull request merged**